### ISSUE TITLE
Port redirecting binding to IPv6 but not IPv4 interfaces.

### ISSUE THREAD

**Issue by marklit** [docker/docker#2174]
Is there a way I can tell docker to only bind redirected ports to IPv4 interfaces?

I have a machine running with IPv6 disabled:

```
# echo '1' > /proc/sys/net/ipv6/conf/lo/disable_ipv6  
# echo '1' > /proc/sys/net/ipv6/conf/lo/disable_ipv6  
# echo '1' > /proc/sys/net/ipv6/conf/all/disable_ipv6  
# echo '1' > /proc/sys/net/ipv6/conf/default/disable_ipv6
# /etc/init.d/networking restart
```

`ifconfig` reports there are no IPv6-enabled interfaces:

```
# ifconfig
docker0   Link encap:Ethernet  HWaddr 00:00:00:00:00:00  
          inet addr:172.17.42.1  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:1372 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7221 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:88091 (88.0 KB)  TX bytes:10655750 (10.6 MB)

eth0      Link encap:Ethernet  HWaddr 04:01:08:c1:b1:01  
          inet addr:198.XXX.XXX.XXX  Bcast:198.199.90.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:97602 errors:0 dropped:4 overruns:0 frame:0
          TX packets:15362 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:141867997 (141.8 MB)  TX bytes:1376970 (1.3 MB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

lxcbr0    Link encap:Ethernet  HWaddr 9e:51:04:ed:13:d4  
          inet addr:10.0.3.1  Bcast:10.0.3.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```

When I launch a new docker container and ask it to port forward 8000 to 8000 in the container it does so only on IPv6 interfaces. Is there a way to make it only bind to IPv4 interfaces?

```
# docker run -p 8000:8000 -i -t colinsurprenant/ubuntu-raring-amd64 /bin/bash
```

When I check with `lsof` it says that only IPv6-related bindings have been made:

```
# lsof -OnP | grep LISTEN
sshd      1275             root    3u     IPv4 ... TCP *:22 (LISTEN)
sshd      1275             root    4u     IPv6 ... TCP *:22 (LISTEN)
dnsmasq   2975      lxc-dnsmasq    7u     IPv4 ... TCP 10.0.3.1:53 (LISTEN)
docker    9629             root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9630        root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9631        root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9632        root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9633        root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9634        root    7u     IPv6 ... TCP *:8000 (LISTEN)
docker    9629 9698        root    7u     IPv6 ... TCP *:8000 (LISTEN)
```

**Comment by m1keil** [comment]
I think there's some kind of misunderstanding of netstat output here. The fact that Proto = tcp6 [doesn't mean](https://blog.codecentric.de/en/2014/04/note-netstat/) that process is bound to IPv6 address only.

**Comment by icoz** [comment]
@m1keil may be there's some kind of misunderstanding of netstat output, but service is unavailable by IPv4.

**Comment by hellais** [comment]
I am also experiencing this bug. Makes me a very sad panda :panda_face:

**Comment by johanhaleby** [comment]
I'm also seeing this bug

**Comment by geekpete** [comment]
Hi Team,

This might have something to do with firewalls, specifically the NAT rules docker dynamically configures for its containers on its local/private network.

If any of you have custom firewall scripts that are flushing/deleting your iptables then replacing them with the rules you want then this might be affecting the proper operation of your docker install.
And if this happens after docker has set up it's own NAT related rules that it needs to communicate with the local 172.17.x.x network it uses, then you'll see it listening on the correct external port but not working since the NAT iptables rule connecting it is simply missing.

One way to test if this is the case is to uninstall/purge docker from your host, a debian/ubuntu example would be:

```
# adjust for your version/package name
apt-get remove lxc-docker-1.2.0 --purge
apt-get autoremove --purge
```

Then reinstall it again and re-run your docker container.

If your app is now working and accessible externally, then your required docker firewall rules have been replaced.

To see them, try:
```
iptables -L -t nat
```
So what you now have to do is ensure your custom firewall script somehow plays nice with the rules docker needs to create/keep.

This approach worked for me anyway, your mileage may vary.

**Comment by laskaridis** [comment]
I'm having the same problem.

1. Started the following container:
```
docker run -d -P --name web training/webapp python app.py
```

2. Verified that container is deployed
```
root@docker:~# docker ps
CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES 
43554e417640 training/webapp:latest "python app.py" 9 minutes ago Up 9 minutes 0.0.0.0:49154->5000/tcp web
```

3. From my server i can access the application in port 49154 as expected:
```
root@docker:~# curl -I localhost:49154 
HTTP/1.0 200 OK 
Content-Type: text/html; charset=utf-8 
Content-Length: 12 
Server: Werkzeug/0.8.3 Python/2.7.3 
Date: Fri, 06 Mar 2015 10:23:24 GMT
```

4. However I cannot access the application in port 49154 from outside the droplet:
```
lefteris@lefteris:~$ curl -I myserver:49154 
curl: (7) couldn't connect to host
```

5. Although i have access to my remote server (public IP removed on purpose):
```
lefteris@lefteris:~$ ping -c3 myserver
PING X.X.X.X (X.X.X.X) 56(84) bytes of data. 
64 bytes from X.X.X.X: icmp_req=1 ttl=52 time=56.3 ms 
64 bytes from X.X.X.X: icmp_req=2 ttl=52 time=56.1 ms 
64 bytes from X.X.X.X: icmp_req=3 ttl=52 time=55.5 ms
```

6. I have verified that my firewall is not blocking my remote request:
```
root@docker:~# iptables -L 
Chain INPUT (policy ACCEPT) 
target prot opt source destination

Chain FORWARD (policy ACCEPT) 
target prot opt source destination 
DOCKER all -- anywhere anywhere 
ACCEPT all -- anywhere anywhere ctstate RELATED,ESTABLISHED 
ACCEPT all -- anywhere anywhere 
ACCEPT all -- anywhere anywhere

Chain OUTPUT (policy ACCEPT) 
target prot opt source destination

Chain DOCKER (1 references) 
target prot opt source destination
```

7. However i can see via netstat that indeed the docker port has been mapped to a ipv6 port in the host sytem, which is the cause of the problem.

**Comment by xinity** [comment]
same problem trying to use collectd in container.
been knocking my head few hours as i saw that the firewall seems to forward udp6 instead of udp4 even if ipv6 is disabled on my host :(

udp        0      0 0.0.0.0:714             0.0.0.0:*                           3507/rpcbind    
udp        0      0 127.0.0.1:764           0.0.0.0:*                           3556/rpc.statd  
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           4674/avahi-daemon: 
udp        0      0 0.0.0.0:58732           0.0.0.0:*                           4674/avahi-daemon: 
udp        0      0 0.0.0.0:34812           0.0.0.0:*                           3556/rpc.statd  
udp        0      0 0.0.0.0:111             0.0.0.0:*                           3507/rpcbind    
udp6       0      0 :::714                  :::*                                3507/rpcbind    
udp6       0      0 :::25826                :::*                                27886/docker-proxy
udp6       0      0 :::5353                 :::*                                4674/avahi-daemon: 
udp6       0      0 :::25968                :::*                                13228/docker-proxy
udp6       0      0 :::42625                :::*                                4674/avahi-daemon: 
udp6       0      0 :::2003                 :::*                                28237/docker-proxy
udp6       0      0 :::34966                :::*                                3556/rpc.statd  
udp6       0      0 :::111                  :::*                                3507/rpcbind

**Comment by riyanshaik** [comment]
Ran into a similar issue on centos 7.Running this would bind dockerd to ipv6
/usr/bin/docker -d -H tcp://0.0.0.0:6969 -H unix:///var/run/docker.sock --insecure-registry=0.0.0.0/0 --ip=0.0.0.0

But this wont,binds to ipv4.
/usr/bin/docker -d -H tcp://10.0.1.10:6969 -H unix:///var/run/docker.sock --insecure-registry=0.0.0.0/0 --ip=0.0.0.0

Disabled ipv6 and it seemed to work.

**Comment by motaro** [comment]
Is this issue fixed. i am having same issue. I have even disabled ipv6 as well but still it keep binding to it

**Comment by ebuildy** [comment]
Still having this issue on Ubuntu 15.04, with Docker version 1.6.2, build 7c8fca2. This is very blocking because Azure load-balancer speaks only IPV4 ;-(

**Comment by seanmcl** [comment]
Same issue with docker 1.6.2 on Centos 7.

**Comment by padcom** [comment]
Same issue with Docker 1.7.0 on Linux Mint 17.1

**Comment by sleaze** [comment]
I've just started encountering this issue within the past week (before that ipv4 binds were not a problem).

Using the EC2 Ubuntu Trusty daily AMI.

> $ sudo docker version
> Client version: 1.7.1
> Client API version: 1.19
> Go version (client): go1.4.2
> Git commit (client): 786b29d
> OS/Arch (client): linux/amd64
> Server version: 1.7.1
> Server API version: 1.19
> Go version (server): go1.4.2
> Git commit (server): 786b29d
> OS/Arch (server): linux/amd64

**Comment by sleaze** [comment]
I just read through this whole thread and [geekpete](https://github.com/docker/docker/issues/2174#issuecomment-77102908) nailed my issue.

I was securing docker containers to only be able to communicate outside my LAN with some `FORWARD` iptables rules.  After removing them, things seem to work.

Very shitty of docker IMO to not able to just explicitly bind to the ipv4 interface and play nice with user iptables rules.

**Comment by chriswayg** [comment]
I have the same issue with Docker 1.8.1 on Debian Jessie. I had to disable IPv6 so I could connect.

**Comment by imesias** [comment]
I wish I had found this thread before I hit this issue, I too discovered with a newly built image, when I deployed a container; docker appeared to only expose the container ports on **ipv6** on the docker host. It was only after much cursing, investigation and caffeine, I found the process inside the container to be binding to **127.0.0.1**. It was after I changed the process to bind to **all interfaces** or **0.0.0.0** inside the container and docker properly exposed the **ipv4** port on the docker host and LAN was able to communicate with the container again.

**Comment by raine** [comment]
I was unable to access [bosun](https://bosun.org)'s HTTP server inside a container created from their `stackexchange/bosun` image.

A solution was to access it through the `docker-machine ip <name>` IP instead of `localhost`.

So, `http://<docker-machine ip dev>:8070` worked.

**Comment by mashuai** [comment]
I have the same issue with docker 1.8.2 on ubuntu 14.04

**Comment by udobranco** [comment]
Hello.

I have this issiu. I need a container with an open tcp port so another container can send it messages (I'm using offical rep rabbitmq image).

Global information: 
I'm using Ubuntu 14.04.2 LTS (GNU/Linux 3.16.0-38-generic x86_64)
Docker version:
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Wed Oct  7 17:48:28 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Wed Oct  7 17:48:28 UTC 2015
 OS/Arch:      linux/amd64

following @bharrisau suggest I've setup ipv6 bind to 1
net.ipv6.bindv6only = 1

Now what I'm running is: 

1)docker run -dt -it -d --hostname my-rabbit --name rabbish-rabbit -e RABBITMQ_ERLANG_COOKIE='secret cookie here' -e RABBIT_MQ_NODENAME=@my-rabbit -p 55555:55555 rabbitmq:3 bash

2)docker run -dt -it -d --hostname my-rabbit --name free-rabbit -e RABBITMQ_ERLANG_COOKIE='secret cookie here' -e RABBIT_MQ_NODENAME=@my-rabbit -P rabbitmq:3 bash

docker ps -a

b5bbeef6657f        rabbitmq:3          (...)           0.0.0.0:32768->5672/tcp              free-rabbit
4230d06ba47e        rabbitmq:3         (...)     5672/tcp, 0.0.0.0:55555->55555/tcp   rabbish-rabbit

telnetting:
1) telnet localhost 55555
Trying ::1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

2)telnet localhost 32768
Trying ::1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

netstat -ntlp

tcp        0      0 0.0.0.0:3306            0.0.0.0:*               LISTEN      1832/mysqld
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1109/sshd
tcp6       0      0 :::4369                 :::*                    LISTEN      1908/epmd
tcp6       0      0 :::22                   :::*                    LISTEN      1109/sshd
tcp6       0      0 :::32768                :::*                    LISTEN      2220/docker-proxy
tcp6       0      0 :::55555                :::*                    LISTEN      2253/docker-proxy

**Comment by cema-sp** [comment]
Hello,
I recently faced this problem and to solve it spent quite a long time, though it was simple: to check on which protocol does application inside container listen to.

**Comment by ktorn** [comment]
I've just encountered this issue when running docker on a Raspberry Pi 2 with Arch Linux.

Starting the container with: `docker run -d -P -p 8080:8080 --name mystuff`
A `docker ps` would show the port was mapped: `0.0.0.0:8080->8080/tcp`
And a `netstat -lnput` would show: `tcp6   0   0   :::8080   :::*   LISTEN   -`

I could access the container at port `8080` from within the Pi, or remotely by port forwarding `ssh  <pi_public_ip> -L8080:localhost:8080`, but not remotely on the Pi's public IP.

Fortunately I found the solution to my problem [here](http://blog.kylemanna.com/linux/2015/07/11/docker--systemd-network-surprise/).

To fix it I simply edited the file `/etc/systemd/network/eth0.network` and added `IPMasquerade=yes` under the `[Network]` section.

**Comment by zx1986** [comment]
I got the same problem now :-(

Ubuntu 14.04.3 LTS
Docker version 1.8.3
Kernel Version: 4.1.5-x86_64-linode61

**Comment by spital** [comment]
it did not work for me ... finally I removed ipv6dis.conf from /etc/modprobe.d and hashed out ipv6* in /etc/sysconfig/network and upgraded docker (probably 1.7 before,not sure)

works now ( https://github.com/sameersbn/docker-gitlab )

# docker version
Client:
 Version:      1.8.2-fc23
 API version:  1.20
 Package Version: docker-1.8.2-7.gitcb216be.fc23.x86_64
 Go version:   go1.5.1
 Git commit:   77d0d91-dirty
 Built:        Mon Oct 12 16:58:35 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2-fc23
 API version:  1.20
 Package Version: 
 Go version:   go1.5.1
 Git commit:   77d0d91-dirty
 Built:        Mon Oct 12 16:58:35 UTC 2015
 OS/Arch:      linux/amd64

Fedora release 23
# uname -r
4.2.3-300.fc23.x86_64
# sysctl net.ipv6.conf.all.forwarding
net.ipv6.conf.all.forwarding = 0
# sysctl net.ipv6.bindv6only
net.ipv6.bindv6only = 0

**Comment by myniva** [comment]
The issue described here is exactly what I could observe in my installation as well. But it turned out to be a different problem: The service in the container did not bind to the external network interface, but only to localhost/127.0.0.1 - with this, no connection could be established, of course.

A comment on Stack Overflow (http://stackoverflow.com/a/30006034/270357) also learned me that sometimes it just *looks* like services do not bind to IPv4.

**Comment by nathanagood** [comment]
In my case, I was using an AWS docker machine and the corporate network where I was working must have been blocking port 2376. I was able to get onto the docker machine using `docker-machine ssh aws01` and verify using netstat -l and ps aux that the docker service was running and listening to port 2376. I was also able to connect to the port using `telnet localhost 2376` as well as `telnet PUBLICIP 2376` so at that point it was reasonable that a firewall was blocking the connection between my laptop and the AWS instance for that port.

**Comment by fredrikaverpil** [comment]
I'm on Docker version 1.10.0-rc2, build c1cdc6e and I'm hitting this very issue.

**Comment by primalmotion** [comment]
same for me with version 1.10 (Toolbox for OS X)

**Comment by ev0lister** [comment]
Same for me here.

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

**Comment by layzerar** [comment]
Got the same issue on `CentOS 7.2.1511` with `Docker version 1.8.2-el7.centos, build a01dc02/1.8.2`.

**I sure there is something wrong with the DNAT rules.**

After I remove the DNAT rules by `iptables -t nat -F DOCKER`, everything works fine. In this situation, the NAT is underlying the `docker-proxy`.

Before I remove the DNAT rules, `telnet 127.0.0.1 1194` is okay, but `telnet 172.16.1.70 1194` was refused. `172.16.1.70` is the LAN address of eth0.

Here is some additional information:
```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS                                            NAMES
dcffd7bd3ecf        fedora-openvpn      "/bin/bash"         About an hour ago   Up 24 minutes       0.0.0.0:1194->1194/tcp, 0.0.0.0:1194->1194/udp   fedora-openvpn
```
```
$ netstat -n -l -t
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
...
tcp6       0      0 :::1194                 :::*                    LISTEN     
tcp6       0      0 :::3306                 :::*                    LISTEN      
```
```
$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: em1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
    link/ether d4:ae:52:a3:fe:84 brd ff:ff:ff:ff:ff:ff
    inet 172.16.1.70/16 brd 172.16.255.255 scope global em1
       valid_lft forever preferred_lft forever
    inet6 fe80::d6ae:52ff:fea3:fe84/64 scope link 
       valid_lft forever preferred_lft forever
3: em2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether d4:ae:52:a3:fe:85 brd ff:ff:ff:ff:ff:ff
4: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:f5:84:f1:ca brd ff:ff:ff:ff:ff:ff
    inet 172.17.42.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:f5ff:fe84:f1ca/64 scope link 
       valid_lft forever preferred_lft forever
12: veth52fd9e8@if11: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP 
    link/ether 8e:31:a9:ee:a9:74 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::8c31:a9ff:feee:a974/64 scope link 
       valid_lft forever preferred_lft forever
```
```
$ iptables --list -n -v -t nat
Chain PREROUTING (policy ACCEPT 121 packets, 21266 bytes)
 pkts bytes target     prot opt in     out     source               destination         
   89  5004 DOCKER     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 61 packets, 3780 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 21 packets, 1260 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DOCKER     all  --  *      *       0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 24 packets, 1440 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  206 12360 MASQUERADE  all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match src-type LOCAL
    0     0 MASQUERADE  all  --  *      !docker0  172.17.0.0/16        0.0.0.0/0           
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.1           172.17.0.1           tcp dpt:1194
    0     0 MASQUERADE  udp  --  *      *       172.17.0.1           172.17.0.1           udp dpt:1194
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.1           172.17.0.1           tcp dpt:1194
    0     0 MASQUERADE  udp  --  *      *       172.17.0.1           172.17.0.1           udp dpt:1194

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination         
    3   180 DNAT       tcp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            tcp dpt:1194 to:172.17.0.1:1194
    0     0 DNAT       udp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            udp dpt:1194 to:172.17.0.1:1194
```

**Comment by layzerar** [comment]
```
$ sysctl -p
kernel.shmmax = 4294967296
net.ipv4.ip_forward = 1
net.ipv4.conf.all.rp_filter = 0
net.ipv4.conf.default.rp_filter = 0
net.ipv4.tcp_fin_timeout = 15
net.ipv4.tcp_keepalive_time = 60
net.ipv6.conf.all.forwarding = 1

$ docker start fedora-openvpn
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS                                            NAMES
dcffd7bd3ecf        fedora-openvpn      "/bin/bash"         2 hours ago         Up 7 seconds        0.0.0.0:1194->1194/tcp, 0.0.0.0:1194->1194/udp   fedora-openvpn

$ telnet 127.0.0.1 1194
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
Connection closed by foreign host.

$ telnet 172.16.1.70 1194
Trying 172.16.1.70...
telnet: connect to address 172.16.1.70: Connection refused

$ iptables -t nat -F DOCKER
$ telnet 172.16.1.70 1194
Trying 172.16.1.70...
Connected to 172.16.1.70.
Escape character is '^]'.
Connection closed by foreign host.
```

**Comment by AxelBroeshart** [comment]
Hi, I had the issue that the deployed webapp from docker example did not show up on http://localhost:<theassignedport>, looking at this post if this had somethng to do with it. Started the container via sudo docker run -d -P training/webapp python app.py and then the app was not approachable in the browser. Telnet to the port works fine.
Solution was to create a Docker group and add the user to this group, so that sudo is no longer necessary. Following the intructions of https://docs.docker.com/engine/installation/linux/ubuntulinux/#create-a-docker-group

sudo usermod -aG docker <user>

**Comment by sandys** [comment]
im hitting this same issue on a stock AWS VM with Debian 8.1 . if I run a stupid webserver using nc/netcat, then netstat shoes it binding on both tcp and tcp6. 

But docker is binding only on tcp6.

**Comment by sandys** [comment]
I figured out that it was not the tcp6 that was the problem but bug #20217 was hitting me. The DOCKER-ISOLATION iptables filter was blocking requests from the outside world. Removing it using "sudo iptables -t filter -F DOCKER-ISOLATION" solved the problem

**Comment by kevinxucs** [comment]
Same for Docker 1.10.3 on ubuntu 14.04

**Comment by marriva** [comment]
Same for Docker 1.9.1  build 78ee77d/1.9.1 on CentOS 7 3.10.0-229.el7.x86_64

**Comment by ivan-exness** [comment]
Same issue:
```
root@d1308:/etc/nginx# uname -a
Linux d1308 3.16.0-46-generic #62~14.04.1-Ubuntu SMP Tue Aug 11 16:27:16 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
root@d1308:/etc/nginx# sysctl net.ipv6.conf.all.forwarding
net.ipv6.conf.all.forwarding = 0
root@d1308:/etc/nginx# sysctl net.ipv6.bindv6only
net.ipv6.bindv6only = 0
root@d1308:/etc/nginx# docker info
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 1.11.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 61
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.16.0-46-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 62.91 GiB
Name: d1308
ID: GWSX:6O3W:AWGV:EJ7L:4AF3:HAQN:I2OM:4KNS:KCDC:CU6K:GHGP:BCSE
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
root@d1308:/etc/nginx# docker version
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64
```

**Comment by alainchiasson** [comment]
As stated by @myniva on Jan 27th - I had to bind my process to 0.0.0.0 INSIDE the container ( A sinatra app w/ WEBrick - 'set :bind, 0.0.0.0' )

**Comment by leodutra** [comment]
Any happy update for a sad brazilian?

**Comment by leodutra** [comment]
Java dudes, you can try this Java IPv4 force:
`java -Xmx1024M -Xms1024M -Djava.net.preferIPv4Stack=true`

**Comment by smougenot** [comment]
If you want your container ports to bind on your ipv4 address, just : 

- edit /etc/sysconfig/docker-network 
  - add DOCKER_NETWORK_OPTIONS=-ip=xx.xx.xx.xx
  - xx.xx.xx.xx being your real ipv4 (and not 0.0.0.0)
- restart docker deamon

works for me on docker 1.9.1

**Comment by leodutra** [comment]
@smougenot Can I pass -ip on `run` or `build`, either?

**Comment by leodutra** [comment]
Dudes, discovered my problem was not the bind... the new ISP modem has no NAT Loopback / Reflection.

Gotta love Brazilian ISPs.

More about NAT Loopback [on Wikipedia](https://en.wikipedia.org/wiki/Network_address_translation#NAT_loopback).

Check this first. In some routers this can be disabled by default, as I read in IT forums.

**Comment by stitchcula** [comment]
Finally, I **unmounted ipv6 module of Linux**(centos7)...and container have been binded to IPv4.
`nano /etc/default/grub`
add ipv6.disable=1 at line 6,like:
`GRUB_CMDLINE_LINUX="ipv6.disable=1 ..."`
then reboot
`grub2-mkconfig -o /boot/grub2/grub.cfg`
`reboot`
ps. `echo '1' > /proc/sys/net/ipv6/conf/all/disable_ipv6` and `-p HOSTIP:PORT:PORT` are useless on my servers.

**Comment by twang2218** [comment]
I got the same issue here. `netstat -tulpn` show the published port only mapped to the `tcp6`, and cannot connect to the container. However, after several try and fail, I found that the port is actually working. I tried to run a new container from docker hub directly:

```
docker run -d -p 8080:80 nginx
```

And the 8080 is binded to tcp6, however, I can access the service via tcp4, there is no problem. Then I realized it's the service inside my container bind to `localhost`, which refused any other connection from outside world, even I did port mapping.

After I changed the service configuration in the container to bind `0.0.0.0` instead, it all works.

This is the mistake I made and I blamed to IPv6 binding issue, hope it will helps.

**Comment by denysvitali** [comment]
I'm having the same issues,
apparently disabling iptables "fixes" the issue.
Really not a classy solution

**Comment by cpuguy83** [comment]
@denysvitali Docker already adds this rule to the PREROUTING chain:
```
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL
```

**Comment by sjkyspa** [comment]
same issue on docker 1.12.0

# docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:00:36 2016
 OS/Arch:      linux/amd64
Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:00:36 2016
 OS/Arch:      linux/amd64

### look at the 53 port


tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      7543/java
tcp        0      0 0.0.0.0:9200            0.0.0.0:*               LISTEN      1762/java
tcp        0      0 0.0.0.0:9300            0.0.0.0:*               LISTEN      1762/java
tcp        0      0 0.0.0.0:5044            0.0.0.0:*               LISTEN      24680/java
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      812/sshd
tcp        0      0 10.45.226.15:5050       0.0.0.0:*               LISTEN      1999/mesos-master
tcp        0      0 10.45.226.15:5051       0.0.0.0:*               LISTEN      2070/mesos-slave
tcp        0      0 10.45.226.15:51134      0.0.0.0:*               LISTEN      7543/java
tcp        0      0 0.0.0.0:2181            0.0.0.0:*               LISTEN      22391/java
tcp        0      0 0.0.0.0:33866           0.0.0.0:*               LISTEN      22391/java
tcp6       0      0 :::8400                 :::*                    LISTEN      23603/consul
tcp6       0      0 :::8500                 :::*                    LISTEN      23603/consul
tcp6       0      0 :::53                   :::*                    LISTEN      1125/docker-proxy
tcp6       0      0 :::8600                 :::*                    LISTEN      23603/consul
tcp6       0      0 :::8300                 :::*                    LISTEN      23603/consul
tcp6       0      0 :::8301                 :::*                    LISTEN      23603/consul
tcp6       0      0 :::8302                 :::*                    LISTEN      23603/consul
udp        0      0 192.168.0.1:123         0.0.0.0:*                           1067/ntpd
udp        0      0 120.27.27.18:123        0.0.0.0:*                           1067/ntpd
udp        0      0 10.45.226.15:123        0.0.0.0:*                           1067/ntpd
udp        0      0 127.0.0.1:123           0.0.0.0:*                           1067/ntpd
udp        0      0 0.0.0.0:123             0.0.0.0:*                           1067/ntpd
udp6       0      0 :::53                   :::*                                1133/docker-proxy
udp6       0      0 :::8301                 :::*                                23603/consul
udp6       0      0 :::8302                 :::*                                23603/consul
udp6       0      0 :::123                  :::*                                1067/ntpd
udp6       0      0 :::8600                 :::*                                23603/consul

**Comment by zx1986** [comment]
I got this just now ....

Docker: 1.12.0
Ubuntu: 16.04 x86_64

**Comment by cpuguy83** [comment]
@zx1986 Please provide extra details. Are you sure the process in the container is actually listening on something other than localhost?

**Comment by zx1986** [comment]
@cpuguy83 Thank you guy83!
I had found that was ok, just a display issue:
http://stackoverflow.com/a/30006034/740004

> When setting up the proxy, Docker requests the loopback address '127.0.0.1', Linux realises this is an address that exists in IPv6 (as ::0) and opens on both (but it is formally an IPv6 socket). When you run netstat it sees this and tells you it is an IPv6 - but it is still listening on IPv4. If you have played with your settings a little, you may have disabled this trick Linux does - by setting net.ipv6.bindv6only = 1.

just like @bharrisau said above.

**Comment by kouhin** [comment]
Exactly the same on docker 1.12!
Our server is a ipv4 only server, but docker 1.12 swarm mode binds published port to IPV6. 

When I try `curl -4 http://localhost`, it says `Connection refused`
but when I try `curl -g -6 http://localhost6`, then the curl hangs up. 

`lsof -i :80` shows the TYPE is IPV6. 

I have resolved this problem by disable ipv6: 

```
sysctl -w net.ipv6.conf.all.disable_ipv6=1
sysctl -w net.ipv6.conf.default.disable_ipv6=1
```
restart docker daemon and re-create service. 

After that, everything goes well!

However, after disabling ipv6, lsof still shows the TYPE is IPV6.

**Comment by cpuguy83** [comment]
@zx1986 this is most likely because your "localhost" is resolving to an IPv6 address.

Swarm services do not get exposed on localhost ( IIRC).

**Comment by kouhin** [comment]
Same issue!
I tried to disable ipv6 and restart docker daemon by sysctl, and it works.
Then I re-enabled the ipv6, it works as well. 
Finally, on other machines, I found the only work is to restart the docker daemon.

**Comment by WayneZhouChina** [comment]
Same issue
[root@xxxxx~]# netstat -ntlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:27017         0.0.0.0:*               LISTEN      934/mongod
tcp        0      0 0.0.0.0:9200            0.0.0.0:*               LISTEN      1318/java
tcp        0      0 0.0.0.0:9300            0.0.0.0:*               LISTEN      1318/java
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      860/sshd
tcp        0      0 0.0.0.0:4440            0.0.0.0:*               LISTEN      963/java
tcp6       0      0 :::9000                 :::*                    LISTEN      3412/docker-proxy

[root@xxxxx ~]# telnet localhost 9000
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

**Comment by cpuguy83** [comment]
@WayneZhouChina This shows that telnet connected to 127.0.0.1 on port 9000, and either docker-proxy has nothing to connect to and subsequently closed the connection or the process in the container rejected it.
Please check your configuration inside the container...

**Comment by arthurlogilab** [comment]
seriously ?

**Comment by leodutra** [comment]
Civilization will solve ISIS before this issue.

Sorry, priorities.

Em 10 de ago de 2016 11:05 AM, "Arthur Lutz" <notifications@github.com>
escreveu:

**Comment by cpuguy83** [comment]
@leodutra please explain your problem.

Also see comments above.
I do not believe there is any actual issue here other than user misconfiguration.

**Comment by ehudkaldor** [comment]
same here.

this is me trying to run an NFS container on ports 110/2048. the host itself is also running an NFS server, and you can see the host binds to both IPv4 and IPv6, but docker only binds to IPv6:
$ sudo netstat -pnl  | grep 11
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN      32686/rpcbind   
tcp        0      0 0.0.0.0:48117           0.0.0.0:*               LISTEN      32713/rpc.mountd
tcp6       0      0 :::110                  :::*                    LISTEN      6175/docker-proxy
tcp6       0      0 :::111                  :::*                    LISTEN      32686/rpcbind   
udp        0      0 0.0.0.0:111             0.0.0.0:*                           32686/rpcbind   
udp6       0      0 :::110                  :::*                                6195/docker-proxy
udp6       0      0 :::111                  :::*                                32686/rpcbind   
unix  2      [ ACC ]     STREAM     LISTENING     13736711 1/init              /var/run/docker.sock

$ sudo netstat -pnl  | grep 204
tcp        0      0 0.0.0.0:2049            0.0.0.0:*               LISTEN      -               
tcp6       0      0 :::2048                 :::*                    LISTEN      6160/docker-proxy
tcp6       0      0 :::2049                 :::*                    LISTEN      -               
udp        0      0 0.0.0.0:2049            0.0.0.0:*                           -               
udp6       0      0 :::2048                 :::*                                6167/docker-proxy
udp6       0      0 :::2049                 :::*                                -

**Comment by geekpete** [comment]
Hi Team,

There seems to be multiple reasons why this happens to different people with different configurations.

It might be nice to come up with a cheat sheet or list of steps to try in order of probability to solve this issue. We can base it on all the confirmed fixes people have found for their own scenarios above. We can gather stats on which problem/solutions people have hit more frequently than others.

This way, anyone new experiencing the problem can go through the troubleshooting flow step by step and hopefully it will help a lot of people even if there are still others who try everything known and still cannot solve it. 

Thoughts?

**Comment by leodutra** [comment]
It is not a user configuration problem when too many users have issues.

It's documentation / design problem.

**Comment by cpuguy83** [comment]
PR's are always welcome. Docs live in this repo (under `/docs`). Would you be willing to make a docs update given your experience with this issue?

**Comment by leodutra** [comment]
I could not solve this.
Docker creates this fake binding and Linux has this fowarding workaround
for IPv4.

This should be documented on RUN or something like that.

I'm not very into Docker organization and don't know exactly where to put
this.

I'll try it later.


*Leo Dutra, **on **Facebook <http://www.facebook.com/leodutra.br>
**and LinkedIn
<https://www.linkedin.com/in/leobr>*

2016-08-19 12:42 GMT-03:00 Brian Goff <notifications@github.com>:

**Comment by stevenlee87** [comment]
same here 

 cat /etc/redhat-release 
CentOS Linux release 7.1.1503 (Core)

docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        
 OS/Arch:      linux/amd64
 Experimental: true

netstat -lntp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name             
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      747/sshd            
tcp6       0      0 :::2375                 :::*                    LISTEN      744/dockerd         
tcp6       0      0 :::22                   :::*                    LISTEN      747/sshd

**Comment by RushOnline** [comment]
[SOLVED] until this bug will be fixed you can specify ipv4 address before port:

~~~bash
... -p 192.168.10.11:8080:80 ...
~~~

or in docker-compose:

~~~yaml
    ports:
      - "192.168.10.11:8080:80"
~~~

**Comment by Starefossen** [comment]
I am facing the same issue Oracle Enterprise Linux 7.

**Comment by JPvRiel** [comment]
> [SOLVED] until this bug will be fixed you can specify ipv4 address before port:

No dice for that workaround with latest ubuntu and docker:
```
$ docker --version
Docker version 1.12.1, build 23cf638
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
```

I tried `-p 172.17.0.1:3128:3128`. It appears to work. The squid conf has `http_port 0.0.0.0:3128`. The host shows its supposedly listening:
```
$ netstat -lnt | grep 3128
tcp        0      0 172.17.0.1:3128         0.0.0.0:*               LISTEN 
```

But any connections from the host fail:
```
$ nc -v 172.17.0.1 3128 | head -n 1
nc: connect to 172.17.0.1 port 3128 (tcp) failed: Connection refused
```

If I don't specify an IP with `-p 3128:3128 `, an IPv6 localhost binding occurs.
```
$ netstat -lnt | grep 3128
tcp6       0      0 :::3128                 :::*                    LISTEN    
```

And it connects okay
```
$ nc -v localhost 3128
Connection to localhost 3128 port [tcp/*] succeeded!
```

Using an alternate test proxy installed on the host and bound to both my libvirt/kvm (virbr0) and docker0 host IPs, I found another interesting docker network issue. 

The `HTTP/1.0 407 Access denied` message below is odd. The proxy had an ACL to allow the `docker0` IP subnet (I double checked it was correct), but for whatever reason, the connection and source routing for `docker0` was via the external IP of the host's eth0 instead of the host IP for docker0. 

I can't understand why kernel source routing is working fine, yet when using docker0 the proxy sees the 
external host IP as connecting
```
$ ip r get 192.168.56.1
local 192.168.56.1 dev lo  src 192.168.56.1 
    cache <local> 
$ ip r get 172.17.0.1
local 172.17.0.1 dev lo  src 172.17.0.1 
    cache <local> 
$ nc -v 192.168.56.1 28080
Connection to 192.168.56.1 28080 port [tcp/*] succeeded!
^C
$ nc -v 172.17.0.1 28080
Connection to 172.17.0.1 28080 port [tcp/*] succeeded!
HTTP/1.0 407 Access denied
Content-Type: text/html

<html><body><h1>Access denied</h1><p>Your request has been declined, 172.20.1.131 is not allowed to connect.</p></body></html>
```

**Comment by RushOnline** [comment]
> No dice for that workaround with latest ubuntu and docker:

~~~bash
$ docker --version
Docker version 1.12.1, build 23cf638

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial

$ sysctl net.ipv6.conf.all.disable_ipv6 
net.ipv6.conf.all.disable_ipv6 = 1
~~~

Work for me.

**Comment by JPvRiel** [comment]
@RushOnline thanks, my bad. It works for me now. Sorry `id10t` error on my end...

After debugging, I found an issue my entrypoint script was hanging checking squid config. Meanwhile, the port mapping on the host interface shows up in netstat, even if the container's process/port isn't really up.

**Comment by mrjana** [comment]
What exact problem are we seeing here? Can somebody clarify?

The way things work is in a host which is dual-stack enabled i.e both IPv4 and IPv6 is enabled if you try to create a wildcard IP socket (i.e INADDR_ANY socket) it will create a `AF_INET6` socket with `[::]` wild card address. The default behavior on linux for that is to listen on all the IPv4 and IPv6 addresses available on the host. This default behavior can be changed on the host by setting `/proc/sys/net/ipv6/bindv6only` to `1`. If not, it is going to listen on addresses from both families. When you run `netstat -tnlp` (or any other variant of that) you will see that socket is bound to an AF_INET6 family because that is what the socket is bound to but as I mentioned before linux will listen on the addresses from both the families. So just seeing that the socket is bound to only an ipv6 family socket should not be a concern to any user. 

It is only a concern if there is an actual problem in reaching your container. If there is no such concern we should close this bug as there is no bug and everything is working as intended. If anybody has any issues please let us know and we can see if we can try to create a separate issue for that.

**Comment by RushOnline** [comment]
@mrjana actually docker-proxy don't listen on ipv4 if you specify wildcard (0.0.0.0).

**Comment by mrjana** [comment]
@RushOnline Are you sure?

```
[ec2-user@ip-172-31-27-46 ~]$ docker run -id -p 9090:80 nginx
594451ba44449a72d39b0bc88b568b3a278cc872bf1de961c241e392ef817772
```

```
[ec2-user@ip-172-31-27-46 ~]$ sudo netstat -tnlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      4475/sshd
tcp        0      0 127.0.0.1:25            0.0.0.0:*               LISTEN      2747/master
tcp6       0      0 :::2377                 :::*                    LISTEN      28350/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      28350/dockerd
tcp6       0      0 :::22                   :::*                    LISTEN      4475/sshd
tcp6       0      0 ::1:25                  :::*                    LISTEN      2747/master
tcp6       0      0 :::9090                 :::*                    LISTEN      29069/docker-proxy
```

```
[ec2-user@ip-172-31-27-46 ~]$ curl 172.31.27.46:9090
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
...
```

**Comment by RushOnline** [comment]
@mrjana not sure for now. 2 weeks before the same sequence  as in you example was result "connection reset", for now is ok.

**Comment by mrjana** [comment]
So what do we do with this issue? Should we close this? I am not sure what more we can get out of this issue. It is pretty clear the issues various folks are having with port publish is probably not related to docker-proxy listening issues as it is actually listening to both ipv4 and ipv6 addresses if the host is dual-stacked. It is probably due to some other issue involving may be iptable rules.

**Comment by RushOnline** [comment]
@mrjana I think there are not one way to get this issue, but if user do all right he will not get it. So I think this is not docker bug, but signal to make more diagnosis messages or improve docs. IMHO - you can close it.

**Comment by xcellardoor** [comment]
Just chiming in that I'm suffering from this as well when I try to run a Gogs container on Fedora. The container was running fine, I upgraded the system including docker, restarted the container and now when I check the system the following is what's happening:

- Port 80 is open and listening from Fedora's perspective, bound to 0.0.0.0 (so any interface).
- When I run 'netstat -tulpn' inside the gogs container itself, it shows that only the IPv6 address is listening. There is no IPv4.

Happy to run tests as people instruct, it's not production or anything it's just a silly git repo I play with.

**Comment by saytik** [comment]
I  have  the  same  problem  on  centos 7.  I  have  disabled  ipv6    via  sysctl.  but
tcp6       0      0 :::8080                 :::*                    LISTEN      12969/docker-proxy  
tcp6       0      0 :::80                   :::*                    LISTEN      12874/docker-proxy

**Comment by inongogo** [comment]
My problem when I encountered this issue was that my firewall was blocking the network that the container was a member of. The firewall had a rule at the bottom of the firewall rules that denied all of the ports and IP addresses that wasn't defined in the list above it. My container was part of the subnet 172.17.0.0/16 so after adding it to the Allow-list in the firewall, all was well.

**Comment by saytik** [comment]
this  fix  the problem:

nano /etc/default/grub
add ipv6.disable=1 at line 6,like:
GRUB_CMDLINE_LINUX="ipv6.disable=1 ..."
grub2-mkconfig -o /boot/grub2/grub.cfg
reboot

**Comment by xcellardoor** [comment]
I believe I have found part of the problem - FirewallD, see https://github.com/docker/docker/issues/27491#event-833578007

**Comment by ricemouse** [comment]
I also have this issue, tried most of the solutions above but it didn't help.

$docker run -p 8080:80 -itd swagger-ui-builder
$netstat -anp|grep 8080
$tcp6       0      0 :::8080                 :::*                    LISTEN      -  

My system information:
$ docker --version
Docker version 1.12.3, build 6b644ec
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial
$ sysctl net.ipv6.conf.all.disable_ipv6 
net.ipv6.conf.all.disable_ipv6 = 1

**Comment by cdstelly** [comment]
On an interactive bash with my container, version 1.12: 
`bash-4.3# netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 127.0.0.11:42821        0.0.0.0:*               LISTEN
tcp        0      0 ::ffff:10.255.0.9:9200  :::*                    LISTEN
tcp        0      0 ::ffff:10.255.0.9:9300  :::*                    LISTEN
udp        0      0 127.0.0.11:54850        0.0.0.0:*
udp        0      0 0.0.0.0:57810           0.0.0.0:*
udp        0      0 0.0.0.0:59623           0.0.0.0:*
udp        0      0 0.0.0.0:37047           0.0.0.0:*
udp        0      0 0.0.0.0:41502           0.0.0.0:*
Active UNIX domain sockets (servers and established)
Proto RefCnt Flags       Type       State         I-Node Path
unix  2      [ ]         STREAM     CONNECTED     171357092
unix  2      [ ]         STREAM     CONNECTED     171365534
bash-4.3# nc -vv ::ffff:10.255.0.9 9200
::ffff:10.255.0.9 ([::ffff:10.255.0.9]:9200) open
^Csent 0, rcvd 0
punt!

bash-4.3# nc -vv localhost 9200
nc: localhost (127.0.0.1:9200): Connection refused
sent 0, rcvd 0
bash-4.3# nc -vv 0.0.0.0 9200
nc: 0.0.0.0 (0.0.0.0:9200): Connection refused
sent 0, rcvd 0
`

All the above about 'It will just work - trust us, netstat is broken' seems to be erroneous at best. 

Is there documentation on this as opposed to wading through github comments?

**Comment by cpuguy83** [comment]
Folks, the service in the container ****must**** be listening on something other than localhost for port forwarding to work.

Please do not post things like "`docker run -p <port>:<port> myCustomimage` doesn't work".
If you have a real, *reproducible* case then we want to see it... but if you have some custom image that no one has access to of course we can't do anything to test it. It's just adding more noise to an already incredibly noisy and ambiguous issue.

**Comment by Tahvok** [comment]
@cpuguy83 please see this:

```
ubuntu@neo4j:~$ docker run -d --publish=7474:7474 --publish=7687:7687 --volume=$HOME/neo4j/data:/data --volume=$HOME/neo4j/logs:/logs --volume=$HOME/neo4j/conf:/conf --ulimit=nofile=40000:40000 --name neo4j neo4j:3.1.1-enterprise
9140aa67d2db320bd3875ff54b807d64e7bac62602efb47dcb299bb84de91b11
ubuntu@neo4j:~$ ss -tupln
Netid State      Recv-Q Send-Q                                                           Local Address:Port                                                                          Peer Address:Port              
udp   UNCONN     0      0                                                                            *:68                                                                                       *:*                  
tcp   LISTEN     0      128                                                                          *:22                                                                                       *:*                  
tcp   LISTEN     0      128                                                                         :::7687                                                                                    :::*                  
tcp   LISTEN     0      128                                                                         :::7474                                                                                    :::*                  
tcp   LISTEN     0      128                                                                         :::22                                                                                      :::*                  

```
Using unmodified official ubuntu cloud image.

Setting net.ipv6.conf.all.forwarding=1 will fix the issue. Didn't try other things, but willing to test anything you might want.

**Comment by greyferret** [comment]
Same problem on Ubuntu 16.04.2
`root@git:~# docker -v
Docker version 1.13.1, build 092cba3
root@git:~# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.2 LTS"
root@git:~# netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:10022           0.0.0.0:*               LISTEN      1385/sshd
tcp6        0      0 :::5000            :::*               LISTEN      16794/docker-proxy
tcp6        0      0 :::5005            :::*               LISTEN      62484/docker-proxy
tcp6        0      0 :::80              :::*               LISTEN      62512/docker-proxy
tcp6        0      0 :::22              :::*               LISTEN      62527/docker-proxy
tcp6        0      0 :::443             :::*               LISTEN      62499/docker-proxy`

Fixed by 
`nano /etc/default/grub
...
GRUB_CMDLINE_LINUX="ipv6.disable=1"
...
update-grub
reboot`

**Comment by geekpete** [comment]
This ticket is a beast. It's one (or a variation of one) symptom, but multiple possible causes. This needs to be definitively addressed in the documentation. 

And are there also some circumstances where people still cannot solve the issue for themselves even with this vast array of possible solutions?

**Comment by mostolog** [comment]
Hi.

CC @cpuguy83 
Think I reproduced this issue with 17.03.0-ce.

```
$docker run --rm --detach --publish 80:80 ubuntu:16.04 tail -f /dev/null
dc469064248406348b0291e31766c66f3cb8a854b157d52f4de0cf3272966f43

$netstat -an | grep 80
tcp6       0      0 :::80                   :::*                    ESCUCHAR   
unix  2      [ ]         DGRAM                    18064    

$ifconfig
docker0   Link encap:Ethernet  direcciónHW XX:XX:XX:XX:XX:XX  
          Direc. inet:172.17.0.1  Difus.:0.0.0.0  Másc:255.255.0.0
          ACTIVO DIFUSIÓN FUNCIONANDO MULTICAST  MTU:1500  Métrica:1
          Paquetes RX:50 errores:0 perdidos:0 overruns:0 frame:0
          Paquetes TX:18 errores:0 perdidos:0 overruns:0 carrier:0
          colisiones:0 long.colaTX:0 
          Bytes RX:2792 (2.7 KB)  TX bytes:900 (900.0 B)
eth0      Link encap:Ethernet  direcciónHW XX:XX:XX:XX:XX:XX  
          Direc. inet:XXX.XXX.XXX.XXX  Difus.:XXX.XXX.XXX.XXX  Másc:255.255.255.240
          ACTIVO DIFUSIÓN FUNCIONANDO MULTICAST  MTU:1500  Métrica:1
          Paquetes RX:5633 errores:0 perdidos:49 overruns:0 frame:0
          Paquetes TX:5018 errores:0 perdidos:0 overruns:0 carrier:0
          colisiones:0 long.colaTX:1000 
          Bytes RX:644526 (644.5 KB)  TX bytes:1219242 (1.2 MB)
lo        Link encap:Bucle local  
          Direc. inet:127.0.0.1  Másc:255.0.0.0
          ACTIVO BUCLE FUNCIONANDO  MTU:65536  Métrica:1
          Paquetes RX:2003 errores:0 perdidos:0 overruns:0 frame:0
          Paquetes TX:2003 errores:0 perdidos:0 overruns:0 carrier:0
          colisiones:0 long.colaTX:1 
          Bytes RX:186799 (186.7 KB)  TX bytes:186799 (186.7 KB)
veth89c76d4 Link encap:Ethernet  direcciónHW 82:ff:f4:42:af:0b  
          ACTIVO DIFUSIÓN FUNCIONANDO MULTICAST  MTU:1500  Métrica:1
          Paquetes RX:14 errores:0 perdidos:0 overruns:0 frame:0
          Paquetes TX:6 errores:0 perdidos:0 overruns:0 carrier:0
          colisiones:0 long.colaTX:0 
          Bytes RX:948 (948.0 B)  TX bytes:300 (300.0 B)
```
IPv6 was previously installed on system, but not anymore (if that could have something to do).


Should I use interface to publish as @xcellardoor suggested?
Should I open a new issue?

Regards

**Comment by cpuguy83** [comment]
@mostolog The host is listening since it runs a proxy process for local traffic (hairpinning in particular) and to occupy the port.
I suspect there is something weird with netstat in this case where ipv6 was enabled at some point and where docker is listening on `0.0.0.0`

**Comment by mostolog** [comment]
@cpuguy83 If interested in fixing this, I could run whatever you may need...

**Comment by cpuguy83** [comment]
@mostolog I don't think there's something to fix in such a case.

1. Assuming there's anything listening on the host side is incorrect. It only does for historical reasons (and the whole hairpinning issue).
2. Issue is with netstat

In the current default config, traffic routing to published ports has can follow two paths depending on if it's local traffic or external traffic.

Let's take the following container:
```
# docker run -d -p 80:80 nginx
```
This yields the following iptables NAT rules:
```
# iptables -t nat -L -v
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DOCKER     all  --  any    any     anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    1    60 DOCKER     all  --  any    any     anywhere            !loopback/8           ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 MASQUERADE  all  --  any    !docker0  172.18.0.0/16        anywhere
    0     0 MASQUERADE  all  --  any    !docker_gwbridge  172.19.0.0/16        anywhere
    0     0 MASQUERADE  tcp  --  any    any     172.18.0.2           172.18.0.2           tcp dpt:http

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  docker0 any     anywhere             anywhere
    0     0 RETURN     all  --  docker_gwbridge any     anywhere             anywhere
    0     0 DNAT       tcp  --  !docker0 any     anywhere             anywhere             tcp dpt:http to:172.18.0.2:80
```

This says, all traffic where the destination is local needs to go the the `DOCKER` chain.

Once it hits the `DOCKER` chain anything coming from `docker0` (the bridge interface) is returned immediately (meaning it skips any further processing in that chain). Since there's nothing else in the `PREROUTING` chain for `docker0` traffic this hits the host network stack.
You'll notice there is a `docker-proxy` process running. This is for handling traffic from the bridge. `docker-proxy` listens on port 80 (the requested host port) for the requested interfaces (usually all interfaces) and proxies all requests to the container IP/port in userspace.
_This is the local traffic handling and for local traffic only_.


You'll notice the `DNAT` rule at the end of the `DOCKER` that says anything incoming on port 80 that does not originate from `docker0` (the bridge interface on the host that Docker uses) will be routed to the container IP's port 80.
_This is for external traffic_. This goes directly to the container and does not hit the host network stack (except to be routed through the bridge interface). Even if you had something listening on 80 on the host it would never hit the service if the source matched this rule.

As you can see, doing a `curl localhost` would not be enough to even test the availability of the service and even if `netstat` correctly reported that `docker-proxy` is listening on tcp4, this would not be enough to determine if things are setup correctly as this is for local traffic only.
Note that technically the usage of docker-proxy could even be paired down further to only forward traffic for `<source ip>==<destination ip> && interface==<bridge name>` (e.g. hairpining), but this makes tons more iptables rules which does incur overhead for incoming packets.
And ideally it wouldn't even exist since the bridge can be setup to enable hairpinning (it's not by default, hence docker-proxy), but there's a pretty nasty kernel bug related to this, so we leave it off by default.

Meanwhile here's my netstat output:
```
# netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp6       0      0 :::2377                 :::*                    LISTEN      7814/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      7814/dockerd
tcp6       0      0 :::80                   :::*                    LISTEN      8503/docker-proxy
udp        0      0 0.0.0.0:4789            0.0.0.0:*                           -
udp6       0      0 :::7946                 :::*                                7814/dockerd
```

TCP6 only... but...
```
# curl 127.0.0.1 # use the actual v4 address to make sure `localhost` doesn't resolve to a v6 address
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
root@e426b2319215:/go/src/github.com/docker/docker# curl 127.0.0.1
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

Even `lsof` reports this as v6 (likely due to the listen address):
```
# lsof -n -i TCP
COMMAND    PID USER   FD   TYPE  DEVICE SIZE/OFF NODE NAME
dockerd   7814 root   13u  IPv6 5955398      0t0  TCP *:2377 (LISTEN)
dockerd   7814 root   14u  IPv4 5955400      0t0  TCP 172.17.0.2:55708->172.17.0.2:2377 (ESTABLISHED)
dockerd   7814 root   18u  IPv6 5955401      0t0  TCP 172.17.0.2:2377->172.17.0.2:55708 (ESTABLISHED)
dockerd   7814 root   21u  IPv6 5955403      0t0  TCP *:7946 (LISTEN)
docker-pr 8503 root    4u  IPv6 5957159      0t0  TCP *:http (LISTEN)
```

I'm going to go ahead and close this issue since there does not seem to be anything to actually fix here.
Thanks all for looking into this.
If forwarding is not working for you I strongly recommend taking a look at the process in the container (and make sure ip forwarding is enabled of course!).

**Comment by softwarevamp** [comment]
seem still bind ipv6 only, netstat -anp | grep docker-proxy | grep LISTEN reports:
    [root@ip-172-31-12-254 centos]# netstat -anp | grep docker-proxy | grep -i listen
    tcp6       0      0 :::6379                 :::*                    LISTEN      2374/docker-proxy
    tcp6       0      0 :::8983                 :::*                    LISTEN      2825/docker-proxy
    tcp6       0      0 :::9092                 :::*                    LISTEN      3160/docker-proxy
    tcp6       0      0 :::2181                 :::*                    LISTEN      2402/docker-proxy

**Comment by perlun** [comment]
@softwarevamp 

> seems still binding on ipv6 only

It usually works correct _even though_ it looks like it's binding on ipv6 only. I have this scenario on one Docker serve right now, it works just fine. On another one it doesn't, but _that's because the service isn't listening on the port inside the container_ in my case. The tcp6 thing is a side track, unfortunately.

@cpuguy83, I suggest the issue be locked to avoid further noise. If people find new problems, a new issue had better be created anyway.

**Comment by thaJeztah** [comment]
Agreed; locking the conversation on this issue.
